File: Library/DataStructures/UnionFind.java
// Shameless copy from Competitive Programming 3
class UnionFind {
    private Vector<Integer> p, rank, setSize;
    private int numSets;

    public UnionFind(int N) {
        p = new Vector<Integer>(N);
        rank = new Vector<Integer>(N);
        setSize = new Vector<Integer>(N);
        numSets = N;
        for (int i = 0; i < N; i++) {
            p.add(i);
            rank.add(0);
            setSize.add(1);
        }
    }

    public int findSet(int i) {
        if (p.get(i) == i) return i;
        else {
            int ret = findSet(p.get(i));
            p.set(i, ret);
            return ret;
        }
    }

    public Boolean isSameSet(int i, int j) {
        return findSet(i) == findSet(j);
    }

    public void unionSet(int i, int j) {
        if (!isSameSet(i, j)) {
            numSets--;
            int x = findSet(i), y = findSet(j);
            // rank is used to keep the tree short
            if (rank.get(x) > rank.get(y)) {
                p.set(y, x);
                setSize.set(x, setSize.get(x) + setSize.get(y));
            } else {
                p.set(x, y);
                setSize.set(y, setSize.get(y) + setSize.get(x));
                if (rank.get(x) == rank.get(y)) rank.set(y, rank.get(y) + 1);
            }
        }
    }

    public int numDisjointSets() {
        return numSets;
    }

    public int sizeOfSet(int i) {
        return setSize.get(findSet(i));
    }
}

File: Library/DataStructures/union_find.cpp
#include <iostream>
#include <vector>
using namespace std;


// Shameless copy from Competitive Programming 3
class UnionFind {                                           
	private:
		vector<int> p, rank, setSize;
		int numSets;
	public:

		UnionFind(int N) {
			setSize.assign(N, 1);
			numSets = N;
			rank.assign(N, 0);
			p.assign(N, 0);
			for (int i = 0; i < N; i++) 
				p[i] = i; 
		}

		int findSet(int i) {
			return (p[i] == i) ? i : (p[i] = findSet(p[i]));
		}

		bool isSameSet(int i, int j) {
			return findSet(i) == findSet(j);
		}

		void unionSet(int i, int j) { 
			if (!isSameSet(i, j)) {
				numSets--; 
				int x = findSet(i);
				int y = findSet(j);

				// rank is used to keep the tree short
				if (rank[x] > rank[y]) { 
					p[y] = x;
					setSize[x] += setSize[y];
				} else {
					p[x] = y;
					setSize[y] += setSize[x];
					if (rank[x] == rank[y]) rank[y]++;
				} 
			} 
		}
		int numDisjointSets() {
			return numSets; 
		}
		int sizeOfSet(int i) { 
			return setSize[findSet(i)]; 
		}
};


int main() {
	return 0;
}
